azavea/loam
# Make dataset operations atomic

## Overview

Specifically, this no longer makes permanent changes to the Emscripten
emulated "file system" inside each web worker, which removes the need to
call "close()" on each individual dataset, and the risk of memory leaks
if someone forgets to do this.

Instead, operations that would result in a new dataset, which at this
point includes ~only GDALWarp and GDALTranslate~ GDALWarp,
GDALTranslate, and GDALRasterize, are stored in an array with their
arguments on the main thread. A Loam Dataset is now a source
plus a list of operations, rather than a pointer to an open GDALDatasetH
inside the web worker's memory.

Besides removing the risk of memory leaks, this also improves
performance when multiple transformations are requested consecutively.

To see why, consider a "tiler" style application, which first uses
GDALWarp to select windows of a GeoTIFF and then GDALTranslate to
convert them to PNGs.

Because tiles are requested faster than they can be processed, under the
old model, the worker will receive the GDALWarp commands first, which
will result in it generating a large number of new datasets, one for
each tile. These datasets will remain in the Worker's memory until they
are closed, which can put memory pressure on the application and slow
down later stages. Next, under the old model, each warped tile will be
converted to a PNG and returned to the main thread, and finally, each
dataset will be closed.

Visually, the practical effect of this is that the user will see a blank
map for some time, and then suddenly all the tiles will fill in all at
once.

Under the new model, each tile is warped, converted to a PNG, and closed
all in one step before processing the next tile, so tiles appear
immediately as they are processed.

One area where this could hurt performance is if a dataset has a long,
complex chain of operations applied to it and then the user wants to
access multiple properties of the resulting dataset. Currently, this
will need to re-calculate the entire dataset each time, which will make
each individual property access slow. The best workaround at this point
for this situation is to call bytes() as a final step and then create a
new dataset from the resulting Blob. However, it would also be feasible
to allow the user to pass in a request for multiple properties to be
satisfied at once, which might be a helpful future enhancement.

## Notes
- This is not an urgent PR, there is no rush to review it.
- There is one notable breaking API change here, which is that `.closeAndReadBytes()` has been removed in favor of a `.bytes()` method on the dataset. This is an easy change to make and I'm intending to release these changes as part of the 1.0.0 release (the library is currently marked `alpha`), so I just went ahead and removed it, but if backwards compatibility is a concern, I'm happy to add it back with a warning to use `.bytes()` in the future.
- There's one other breaking change, which is the removal of `flushFS()`, but that was primarily written for testing purposes so I think it's fine to just remove it.
- @alkamin I'm planning to take over and finish off #26 as my next task unless you'd still like to work on it--let me know if that's okay.

## Testing Instructions

 * Ensure that CI passes
 * If you want to get major bonus points, you could build the library, and upgrade it in a place where we use Loam in a project (Warpr or RF) to confirm that it still works.

## Checklist


Resolves #21
